llvm/lib/Transforms/Vectorize/LoopVectorize.cpp

```diff
@@ -4813,16 +4839,19 @@ LoopVectorizationCostModel::selectInterleaveCount(VPlan &Plan, ElementCount VF,
       MaxInterleaveCount = ForceTargetMaxVectorInterleaveFactor;
   }
 
-  unsigned EstimatedVF = getEstimatedRuntimeVF(VF, VScaleForTuning);
-
   // Try to get the exact trip count, or an estimate based on profiling data or
-  // ConstantMax from PSE, failing that.
-  if (auto BestKnownTC = getSmallBestKnownTC(PSE, TheLoop)) {
+  // ConstantMax from PSE, failing that. For fixed length VFs treat a scalable
+  // trip count as if unknown.
+  if (auto BestKnownTC = getSmallBestKnownTC(PSE, TheLoop);
+      BestKnownTC && (BestKnownTC->isFixed() || VF.isScalable())) {
+    // Re-evaluate trip counts and VFs to be in the same numerical space.
+    unsigned AvailableTC = getEstimatedRuntimeVF(*BestKnownTC, VScaleForTuning);
```

Comment:
`getEstimatedRuntimeVF` should probably be renamed, using it here to compute `AvailableTC` seems a bit confusing.